### ISSUE TITLE
Cannot build 610_CSGTree tutorial

### ISSUE THREAD

**Issue by Aidoru** [libigl/libigl#177]
Hello,
cmake doesn't generate 610_CSGTree project on Windows.
I don't see any error.

Anyone got problem to build 610 on windows?

Thank you

**Comment by alecjacobson** [comment]
Do you have CGAL installed?

**Comment by Aidoru** [comment]
Sorry, I didn't write right and now I can see error:
yes I have CGAL installed, my problem is that I have that error when I try to build in Visual Studio:

"Error	C2338	YOU_MIXED_DIFFERENT_NUMERIC_TYPES__YOU_NEED_TO_USE_THE_CAST_METHOD_OF_MATRIXBASE_TO_CAST_NUMERIC_TYPES_EXPLICITLY"	

I use Visual Studio 2015.

Thank you

**Comment by alecjacobson** [comment]
Oh, strange. Could you please provide the full compiler error (with line
numbers)?

**Comment by Aidoru** [comment]
I will past all my output:

1>  cl /c /IG:\GIT\libigl\external\nanogui\ext\glew\include /IG:\GIT\libigl\external\nanogui\ext\eigen /IG:\GIT\libigl\external\nanogui\ext\glfw\include /I\usr\local\include /I\opt\local\include /IG:\GIT\libigl\tutorial /IG:\GIT\libigl\tutorial\..\include /I"C:\Program Files\CGAL-4.7\auxiliary\gmp\include" /IG:\GIT\libigl\tutorial\build\610_CSGTree /I"C:\Program Files\CGAL-4.7\build\include" /I"C:\Program Files\CGAL-4.7\include" /IC:\local\boost_1_60_0 /Zi /W3 /WX- /MP /Od /Ob0 /D WIN32 /D _WINDOWS /D _CRT_SECURE_NO_DEPRECATE /D _SCL_SECURE_NO_DEPRECATE /D _CRT_SECURE_NO_WARNINGS /D _SCL_SECURE_NO_WARNINGS /D TETLIBRARY /D TRILIBRARY /D ANSI_DECLARATORS /D _DEBUG /D NO_TIMER /D IGL_OPENGL_4 /D IGL_NO_OPENGL /D "TUTORIAL_SHARED_PATH=\"G:/GIT/libigl/tutorial/shared\"" /D CGAL_USE_MPFR /D CGAL_USE_GMP /D BOOST_ALL_DYN_LINK /D "CMAKE_INTDIR=\"Debug\"" /D _MBCS /Gm- /EHsc /RTC1 /MDd /GS /fp:strict /fp:except- /Zc:wchar_t /Zc:forScope /Zc:inline /GR /Fo"610_CSGTree_bin.dir\Debug\\" /Fd"610_CSGTree_bin.dir\Debug\vc140.pdb" /Gd /TP /wd4503 /errorReport:prompt  /bigobj /bigobj /w G:\GIT\libigl\tutorial\610_CSGTree\main.cpp
1>cl : Command line warning D9025: overriding '/W3' with '/w'
1>
1>  main.cpp
1>g:\git\libigl\external\nanogui\ext\eigen\eigen\src/Core/Assign.h(493): error C2338: YOU_MIXED_DIFFERENT_NUMERIC_TYPES__YOU_NEED_TO_USE_THE_CAST_METHOD_OF_MATRIXBASE_TO_CAST_NUMERIC_TYPES_EXPLICITLY
1>  g:\git\libigl\external\nanogui\ext\eigen\eigen\src/Core/PlainObjectBase.h(414): note: see reference to function template instantiation 'Derived &Eigen::DenseBase<Derived>::lazyAssign<Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::ArrayWrapper<const Eigen::Matrix<igl::copyleft::boolean::CSGTree::FIndex,-1,1,0,-1,1>>>>>(const Eigen::DenseBase<Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::ArrayWrapper<const Eigen::Matrix<igl::copyleft::boolean::CSGTree::FIndex,-1,1,0,-1,1>>>>> &)' being compiled
1>          with
1>          [
1>              Derived=Eigen::Matrix<long,-1,1,0,-1,1>
1>          ]
1>  g:\git\libigl\external\nanogui\ext\eigen\eigen\src/Core/PlainObjectBase.h(414): note: see reference to function template instantiation 'Derived &Eigen::DenseBase<Derived>::lazyAssign<Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::ArrayWrapper<const Eigen::Matrix<igl::copyleft::boolean::CSGTree::FIndex,-1,1,0,-1,1>>>>>(const Eigen::DenseBase<Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::ArrayWrapper<const Eigen::Matrix<igl::copyleft::boolean::CSGTree::FIndex,-1,1,0,-1,1>>>>> &)' being compiled
1>          with
1>          [
1>              Derived=Eigen::Matrix<long,-1,1,0,-1,1>
1>          ]
1>  g:\git\libigl\external\nanogui\ext\eigen\eigen\src/Core/Assign.h(520): note: see reference to function template instantiation 'Derived &Eigen::PlainObjectBase<Derived>::lazyAssign<Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::ArrayWrapper<const Eigen::Matrix<igl::copyleft::boolean::CSGTree::FIndex,-1,1,0,-1,1>>>>>(const Eigen::DenseBase<Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::ArrayWrapper<const Eigen::Matrix<igl::copyleft::boolean::CSGTree::FIndex,-1,1,0,-1,1>>>>> &)' being compiled
1>          with
1>          [
1>              Derived=Eigen::Matrix<long,-1,1,0,-1,1>
1>          ]
1>  g:\git\libigl\external\nanogui\ext\eigen\eigen\src/Core/Assign.h(520): note: see reference to function template instantiation 'Derived &Eigen::PlainObjectBase<Derived>::lazyAssign<Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::ArrayWrapper<const Eigen::Matrix<igl::copyleft::boolean::CSGTree::FIndex,-1,1,0,-1,1>>>>>(const Eigen::DenseBase<Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::ArrayWrapper<const Eigen::Matrix<igl::copyleft::boolean::CSGTree::FIndex,-1,1,0,-1,1>>>>> &)' being compiled
1>          with
1>          [
1>              Derived=Eigen::Matrix<long,-1,1,0,-1,1>
1>          ]
1>  g:\git\libigl\external\nanogui\ext\eigen\eigen\src/Core/Assign.h(520): note: while compiling class template member function 'Eigen::Matrix<long,-1,1,0,-1,1> &Eigen::internal::assign_selector<Derived,OtherDerived,false,false>::run(Derived &,const OtherDerived &)'
1>          with
1>          [
1>              Derived=Eigen::Matrix<long,-1,1,0,-1,1>,
1>              OtherDerived=Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::ArrayWrapper<const Eigen::Matrix<igl::copyleft::boolean::CSGTree::FIndex,-1,1,0,-1,1>>>>
1>          ]
1>  g:\git\libigl\external\nanogui\ext\eigen\eigen\src/Core/Assign.h(564): note: see reference to function template instantiation 'Eigen::Matrix<long,-1,1,0,-1,1> &Eigen::internal::assign_selector<Derived,OtherDerived,false,false>::run(Derived &,const OtherDerived &)' being compiled
1>          with
1>          [
1>              Derived=Eigen::Matrix<long,-1,1,0,-1,1>,
1>              OtherDerived=Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::ArrayWrapper<const Eigen::Matrix<igl::copyleft::boolean::CSGTree::FIndex,-1,1,0,-1,1>>>>
1>          ]
1>  g:\git\libigl\external\nanogui\ext\eigen\eigen\src/Core/Assign.h(564): note: see reference to class template instantiation 'Eigen::internal::assign_selector<Derived,OtherDerived,false,false>' being compiled
1>          with
1>          [
1>              Derived=Eigen::Matrix<long,-1,1,0,-1,1>,
1>              OtherDerived=Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::ArrayWrapper<const Eigen::Matrix<igl::copyleft::boolean::CSGTree::FIndex,-1,1,0,-1,1>>>>
1>          ]
1>  g:\git\libigl\external\nanogui\ext\eigen\eigen\src/Core/PlainObjectBase.h(453): note: see reference to function template instantiation 'Derived &Eigen::MatrixBase<Derived>::operator =<Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::ArrayWrapper<const Eigen::Matrix<igl::copyleft::boolean::CSGTree::FIndex,-1,1,0,-1,1>>>>>(const Eigen::DenseBase<Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::ArrayWrapper<const Eigen::Matrix<igl::copyleft::boolean::CSGTree::FIndex,-1,1,0,-1,1>>>>> &)' being compiled
1>          with
1>          [
1>              Derived=Eigen::Matrix<long,-1,1,0,-1,1>
1>          ]
1>  g:\git\libigl\external\nanogui\ext\eigen\eigen\src/Core/PlainObjectBase.h(453): note: see reference to function template instantiation 'Derived &Eigen::MatrixBase<Derived>::operator =<Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::ArrayWrapper<const Eigen::Matrix<igl::copyleft::boolean::CSGTree::FIndex,-1,1,0,-1,1>>>>>(const Eigen::DenseBase<Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::ArrayWrapper<const Eigen::Matrix<igl::copyleft::boolean::CSGTree::FIndex,-1,1,0,-1,1>>>>> &)' being compiled
1>          with
1>          [
1>              Derived=Eigen::Matrix<long,-1,1,0,-1,1>
1>          ]
1>  g:\git\libigl\external\nanogui\ext\eigen\eigen\src/Core/Matrix.h(184): note: see reference to function template instantiation 'Derived &Eigen::PlainObjectBase<Derived>::operator =<Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::ArrayWrapper<const Eigen::Matrix<igl::copyleft::boolean::CSGTree::FIndex,-1,1,0,-1,1>>>>>(const Eigen::EigenBase<Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::ArrayWrapper<const Eigen::Matrix<igl::copyleft::boolean::CSGTree::FIndex,-1,1,0,-1,1>>>>> &)' being compiled
1>          with
1>          [
1>              Derived=Eigen::Matrix<long,-1,1,0,-1,1>
1>          ]
1>  g:\git\libigl\external\nanogui\ext\eigen\eigen\src/Core/Matrix.h(184): note: see reference to function template instantiation 'Derived &Eigen::PlainObjectBase<Derived>::operator =<Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::ArrayWrapper<const Eigen::Matrix<igl::copyleft::boolean::CSGTree::FIndex,-1,1,0,-1,1>>>>>(const Eigen::EigenBase<Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::ArrayWrapper<const Eigen::Matrix<igl::copyleft::boolean::CSGTree::FIndex,-1,1,0,-1,1>>>>> &)' being compiled
1>          with
1>          [
1>              Derived=Eigen::Matrix<long,-1,1,0,-1,1>
1>          ]
1>  G:\GIT\libigl\tutorial\610_CSGTree\main.cpp(73): note: see reference to function template instantiation 'Eigen::Matrix<long,-1,1,0,-1,1> &Eigen::Matrix<long,-1,1,0,-1,1>::operator =<Derived>(const Eigen::EigenBase<Derived> &)' being compiled
1>          with
1>          [
1>              Derived=Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::ArrayWrapper<const Eigen::Matrix<igl::copyleft::boolean::CSGTree::FIndex,-1,1,0,-1,1>>>>
1>          ]
1>  G:\GIT\libigl\tutorial\610_CSGTree\main.cpp(73): note: see reference to function template instantiation 'Eigen::Matrix<long,-1,1,0,-1,1> &Eigen::Matrix<long,-1,1,0,-1,1>::operator =<Derived>(const Eigen::EigenBase<Derived> &)' being compiled
1>          with
1>          [
1>              Derived=Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::CwiseUnaryOp<Eigen::internal::scalar_add_op<__int64>,const Eigen::ArrayWrapper<const Eigen::Matrix<igl::copyleft::boolean::CSGTree::FIndex,-1,1,0,-1,1>>>>
1>          ]
========== Build: 0 succeeded, 1 failed, 3 up-to-date, 0 skipped ==========


Hope you can understand.

Thank you

**Comment by Aidoru** [comment]
Forgot to tell you that I compiled with Visual Studio 2015 64bit version

**Comment by alecjacobson** [comment]
Ah funny. It's a bit hard to parse, but I'm guess that on your build
`MatrixXi::Index` is not a typedef/template for  `long int` . I've pushed
what's hopefully a fix to that `610.../main.cpp` file.

**Comment by Aidoru** [comment]
Thank you,
I can now build without error, but I have problems on runtime:
I can switch between "view_id" from 0 to 4 (set view_id = 0), when I try "view_id" from 5 to 8 I get error:

A popup tell me that 610_CSGTree_bin.exe has trigger a breakpoint (cursor at the end of "inline void aligned_free(void *ptr) function in file Memory.h)

Then, if I continue, I get a "Debug Assertion Failed":
Program: ..........\.\.\.610_CSGTree_bin.exe
File: minkernel\crts\src\appcrt\heap\debug_heap.cpp
Line: 888
Expression: _CrtlsValidHeapPointer(block)

If I try to continue I enter in a loop of "Debug Assertion Failed"....

**Comment by alecjacobson** [comment]
Could you paste a full backtrace report running this in debug mode? I'd
like to see where in our code this could be coming from.

**Comment by Aidoru** [comment]
This is my call stack when I reach view_id == 5

 	ntdll.dll!00007ffe36d5fab5()	Unknown
 	ntdll.dll!00007ffe36d310a7()	Unknown
 	ntdll.dll!00007ffe36ce5472()	Unknown
 	KernelBase.dll!00007ffe335b464a()	Unknown
 	ucrtbased.dll!00007ffe0b03dd21()	Unknown
 	ucrtbased.dll!00007ffe0b03c0c9()	Unknown
 	ucrtbased.dll!00007ffe0b03f715()	Unknown
 	ucrtbased.dll!00007ffe0b03fdf8()	Unknown
>	610_CSGTree_bin.exe!Eigen::internal::aligned_free(void * ptr) Line 253	C++
 	610_CSGTree_bin.exe!Eigen::internal::conditional_aligned_free<1>(void * ptr) Line 318	C++
 	610_CSGTree_bin.exe!Eigen::internal::conditional_aligned_delete_auto<double,1>(double * ptr, unsigned __int64 size) Line 446	C++
 	610_CSGTree_bin.exe!Eigen::DenseStorage<double,-1,-1,-1,0>::~DenseStorage<double,-1,-1,-1,0>() Line 236	C++
 	[External Code]	
 	610_CSGTree_bin.exe!main::__l2::<lambda>() Line 89	C++
 	610_CSGTree_bin.exe!main::__l2::<lambda>(igl::viewer::Viewer & viewer, unsigned char key, int mods) Line 125	C++
 	[External Code]	
 	610_CSGTree_bin.exe!igl::viewer::Viewer::key_down(int key, int modifiers) Line 466	C++
 	610_CSGTree_bin.exe!glfw_key_callback(GLFWwindow * window, int key, int scancode, int action, int modifier) Line 142	C++
 	610_CSGTree_bin.exe!_glfwInputKey(_GLFWwindow * window, int key, int scancode, int action, int mods) Line 157	C
 	610_CSGTree_bin.exe!windowProc(HWND__ * hWnd, unsigned int uMsg, unsigned __int64 wParam, __int64 lParam) Line 313	C
 	[External Code]	
 	610_CSGTree_bin.exe!_glfwPlatformPollEvents() Line 1014	C
 	610_CSGTree_bin.exe!_glfwPlatformWaitEvents() Line 1053	C
 	610_CSGTree_bin.exe!glfwWaitEvents() Line 729	C
 	610_CSGTree_bin.exe!igl::viewer::Viewer::launch_rendering(bool loop) Line 897	C++
 	610_CSGTree_bin.exe!igl::viewer::Viewer::launch(bool resizable, bool fullscreen) Line 926	C++
 	610_CSGTree_bin.exe!main(int argc, char * * argv) Line 127	C++
 	[External Code]

**Comment by alecjacobson** [comment]
Is it possible to compile in debug mode so that the backtrace reveals which
files are invoking the error?

**Comment by Aidoru** [comment]
I will describe step by step what happened:

1) I commented the line "#warning "Constructing Eigen::PlainObjectBase directly is deprecated" in   slice.cpp, because I get error "invalid preprocessor command 'warning'

2) In main.cpp I changed "int view_id = num_views-1;" to "int view_id = 0"

3) Running in debug mode, as I tell you, I can switch from view_id = 0 to view_id = 4

4) When I reach view_id == 5, if I go on "step into", on line " M = {{VA,FA},{VB,FB},"i"};" in main.cpp,
program stop when I jump from line "viewer.data.set_mesh(M.cast_V<MatrixXd>(),M.F());" in main.cpp, to "inline void aligned_free(void *ptr)" in Memory.h (not the first call if you step into, but after 25 call)

**Comment by wguo68** [comment]
Is it due to the fact:  the CGAL does not support VS2015 now?

**Comment by Aidoru** [comment]
Didn't remember any errors on compiling CGAL for VS2015

**Comment by wguo68** [comment]
How do you compile CGAL for VS2015?

**Comment by Aidoru** [comment]
CMake

**Comment by alecjacobson** [comment]
Aidoru, did you ever resolve this problem?

**Comment by Aidoru** [comment]
Ops, sorry I've see now your question. 
Unfortunately nothing to do
